## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

In [204]:
arr = np.random.randint(0,10,10)

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [20]:
to_replace = (arr >= 3) & (arr <= 8)
arr[to_replace] = np.negative(arr[to_replace])
arr

array([ 9,  0,  2, -5,  1, -4,  2,  0, -6,  0], dtype=int32)

**2.** Заменить максимальный элемент случайного массива на 0

In [22]:
arr[arr==max(arr)] = 0
arr

array([ 0,  0,  2, -5,  1, -4,  2,  0, -6,  0], dtype=int32)

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [206]:
arr_2dim = np.random.randint(0,10,(2,5))

In [47]:
np.array(np.meshgrid(arr_2dim[0],arr_2dim[1])).T.reshape(-1,2)


array([[1, 3],
       [1, 1],
       [1, 2],
       [1, 0],
       [1, 4],
       [1, 3],
       [1, 1],
       [1, 2],
       [1, 0],
       [1, 4],
       [3, 3],
       [3, 1],
       [3, 2],
       [3, 0],
       [3, 4],
       [7, 3],
       [7, 1],
       [7, 2],
       [7, 0],
       [7, 4],
       [9, 3],
       [9, 1],
       [9, 2],
       [9, 0],
       [9, 4]], dtype=int32)

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [61]:
A = np.random.randint(0,10,(8,3))
B = np.random.randint(0,10,(2,2))

In [207]:
A

array([[7, 7, 2],
       [2, 7, 4],
       [5, 6, 1],
       [9, 6, 1],
       [4, 9, 6],
       [2, 7, 7],
       [7, 5, 9],
       [1, 5, 0]], dtype=int32)

In [63]:
B

array([[9, 2],
       [7, 4]], dtype=int32)

In [176]:
A[np.bool(np.sum(np.isin(A,B[0]),axis=1) * np.sum(np.isin(A,B[1]),axis=1))]

array([[7, 7, 2],
       [2, 7, 4],
       [4, 9, 6],
       [2, 7, 7],
       [7, 5, 9]], dtype=int32)

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [78]:
matrix = np.random.randint(1,3,(10,3))

In [79]:
matrix

array([[2, 2, 2],
       [1, 2, 1],
       [2, 1, 1],
       [2, 2, 2],
       [2, 2, 2],
       [2, 1, 2],
       [1, 1, 1],
       [2, 1, 1],
       [2, 1, 2],
       [1, 2, 1]], dtype=int32)

In [82]:
matrix[np.bool(matrix.max(axis=1)-matrix.min(axis=1))]

array([[1, 2, 1],
       [2, 1, 1],
       [2, 1, 2],
       [2, 1, 1],
       [2, 1, 2],
       [1, 2, 1]], dtype=int32)

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [3]:
arr_2 = np.random.randint(0,3,(10,3))
arr_2

array([[2, 2, 0],
       [0, 2, 0],
       [0, 2, 1],
       [2, 0, 0],
       [0, 2, 1],
       [2, 2, 0],
       [2, 0, 1],
       [0, 2, 2],
       [2, 0, 2],
       [1, 2, 2]], dtype=int32)

In [9]:
np.unique(arr_2,axis=0)

array([[0, 2, 0],
       [0, 2, 1],
       [0, 2, 2],
       [1, 2, 2],
       [2, 0, 0],
       [2, 0, 1],
       [2, 0, 2],
       [2, 2, 0]], dtype=int32)

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [214]:
k = 1
for i,line in enumerate(arr_2dim):
    k*=line[i]
k

14

In [221]:
np.prod(arr_2dim.diagonal())

14

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [270]:
v1=np.array([1, 2, 4])
v2 = np.array([4, 2, 1, 2])

In [413]:
sorted(v1) == sorted(v2)

In [271]:
np.array_equal(np.sort(v1),np.sort(v2))

False

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [304]:
arr_3_np = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

In [275]:
arr_3 = [6, 2, 0, 3, 0, 0, 5, 7, 0]

In [276]:
first_index = arr_3.index(0)+1
max_val = arr[first_index]
for i in range(first_index+1,len(arr_3)):
    if arr_3[i-1] == 0 and arr_3[i]>max_val:
        max_val = arr_3[i]
max_val

5

In [305]:
np.max(np.take(arr_3_np,np.subtract(np.where(arr_3_np==0),-1),mode='clip'))

np.int64(5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [307]:
arr_4 = [2, 2, 2, 3, 3, 3, 5]
arr_4_np = np.array([2, 2, 2, 3, 3, 3, 5])

In [315]:
res_dict = dict()
for each in arr_4:
    if each in res_dict:
        res_dict[each]+=1
    else:
        res_dict[each]=1 
(list(res_dict.keys()),list(res_dict.values()))

([2, 3, 5], [3, 3, 1])

In [316]:
np.unique(arr_4_np,return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [39]:
X = np.random.randint(0,10,(1,5))
Y = np.random.randint(0,10,(1,5))
x_list = X.tolist()[0]
y_list = Y.tolist()[0]

In [42]:
res = list()
for i, y in enumerate(y_list):
    res.append([])
    for x in x_list:
        res[i].append(abs(x-y))
res

[[0, 6, 4, 3, 0],
 [1, 7, 5, 4, 1],
 [7, 1, 3, 4, 7],
 [7, 1, 3, 4, 7],
 [4, 2, 0, 1, 4]]

In [41]:
m, n = np.meshgrid(X, Y)
abs(m-n)

array([[0, 6, 4, 3, 0],
       [1, 7, 5, 4, 1],
       [7, 1, 3, 4, 7],
       [7, 1, 3, 4, 7],
       [4, 2, 0, 1, 4]], dtype=int32)

In [ ]:
import time
import scipy

In [412]:
start_time = time.time()
for _ in range(0,10000):
    m, n = np.meshgrid(X, Y)
    abs(m-n)
time.time() - start_time

1.8498773574829102

In [410]:
start_time = time.time()
for _ in range(0,10000):
    scipy.spatial.distance.cdist(X,Y,'euclidean')
round(time.time() - start_time, 10)

0.3015959263

_______
________

При увеличении количества запусков программы скорость работы будет изменяться еще больше в пользу функции из библиотеки scipy

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [77]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [58]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [434]:
np.average(calorie_stats)-60

np.float64(46.883116883116884)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [436]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [437]:
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [71]:
proc = 2
while True:
    calories = np.percentile(calorie_stats,proc)
    print(round(proc,2),round(calories,2))
    if calories > 60:
        nth_percentile = proc
        break
    proc+=0.1
round(nth_percentile,2)

2 50.0
2.1 50.0
2.2 50.0
2.3 50.0
2.4 50.0
2.5 50.0
2.6 50.0
2.7 51.04
2.8 52.56
2.9 54.08
3.0 55.6
3.1 57.12
3.2 58.64
3.3 60.16


3.3

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [73]:
more_calories = calorie_stats[calorie_stats>60].size / calorie_stats.size *100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [456]:
calorie_std = np.std(calorie_stats)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

 В результате можно сделать вывод, что компания CrunchieMunchies выпускает довольно качественную продукцию. Среднее значение калорий в продукции меньше, чем у конкурентов на 46 калорий, при этом доля хлопьев, в которых количество калорий больше, чем у нашей продукции, составляет 96% среди конкурентов. Так как стандартное отклонение равно 19, можно сделать вывод, что значения калорий у конкурентов в основном варьируются в диапазоне от 87 до 125 калорий. Это можно использовать при продвижении компании на мировом рынке.